### ISSUE TITLE
java: ClassDict should extend LinkedHashMap in order to provide predictable attribute iteration order?

### ISSUE THREAD

**Issue by vruusmann** [irmen/Pyrolite#37]
Do you have any information if Python makes any guarantees about the ordering of `dict` attributes (eg. declaration order, initialization order, etc)? If it does, then it should be reflected at the `ClassDict` level by choosing a more specialized parent class.

Of course, there will be a slight performance penalty (perhaps 10 to 15%) when switching from `HashMap` to `LinkedHashMap`.

**Comment by vruusmann** [comment]
This question can be generalized to all `dict`-type attributes.

For example, it appears to be the case that `numpy.dtype.values` should have a predictable attribute iteration order. The `HashMap` implementation class does not reflect it, which breaks the unpickling behaviour for `sklearn.tree._tree.tree` class (I need to rewrite the `value` attribute in method `DType#toDescr()`): https://github.com/jpmml/jpmml-sklearn/blob/master/src/main/java/numpy/DType.java

Or perhaps it only applies to CPython classes.

**Comment by irmen** [comment]
Dictionaries in Python don't have a guaranteed order of their values. What's more, there's even a setting that allows you to control/change the hash algorithm's seed value (PYTHONHASHSEED) which in turn will change the apparent order of keys in a dict.

If you're using OrderedDict however, the order of the keys *is* guaranteed to be the same as the order in which they are inserted in the dict https://docs.python.org/3/library/collections.html#collections.OrderedDict

I don't know what numpy.dtype.values is and a quick google didn't teach me much. What does python itself do when you pickle/unpickle such a value?

**Comment by vruusmann** [comment]
OK, you have clearly thought about it before.

I have a suspicion that CPython classes assume fixed/positional attribute ordering. The CPython object initialization method takes an array of arguments (eg. `__init__(Object[])`), and it is the responsibility of the CPython class to establish the mapping between its attributes and array elements.

In that case, my `CClassDict` should extend `java.util.LinkedHashMap` (or provide an alternative mechanism for communicating the list of attribute names), whereas the regular `ClassDict` can stay a subclass of `java.util.HashMap`.

**Comment by irmen** [comment]
Somehow I have the impression that you're confusing parts of object creation using the ``__init__`` method and its parameter list, an object's attributes, and filling a dictionary with key/value pairs. Those are all different things. Recreating a dict via pickle is doing peculiar things too. Doesn't matter, you seem to have solved your issue.
